src/plugin.ts

```diff
@@ -0,0 +1,212 @@
+import { isInsideIFrame } from './utils/iframe';
+import { buildElementLink } from './utils/link';
+import {
+  IElementClickMessageData,
+  IFrameCommunicator,
+  IFrameMessageType,
+  IHighlightsStatusMessageData,
+} from './lib/IFrameCommunicator';
+import { NodeHighlighter, NodeHighlighterEventType } from './lib/NodeHighlighter';
+import { createStorage } from './utils/storage';
+
+interface IPluginState {
+  readonly queryTimerId: ReturnType<typeof setTimeout>;
+}
+
+interface IPluginIFrameSettings {
+  readonly highlighterEnabled: boolean;
+}
+
+export interface IPluginConfiguration {
+  readonly languageCodename: string | null;
+  readonly projectId: string | null;
+  readonly queryParam: string;
+}
+
+const initialState: IPluginState = {
+  queryTimerId: 0,
+};
+
+const defaultConfiguration: IPluginConfiguration = {
+  languageCodename: null,
+  projectId: null,
+  queryParam: 'kk-plugin-enabled',
+};
+
+class Plugin {
+  private state: IPluginState;
+  private configuration: IPluginConfiguration;
+  private iFrameCommunicator?: IFrameCommunicator;
+  private readonly nodeHighlighter: NodeHighlighter;
+
+  constructor(configuration?: Partial<IPluginConfiguration>) {
+    this.nodeHighlighter = new NodeHighlighter();
+
+    this.state = initialState;
+    this.configuration = { ...defaultConfiguration, ...configuration };
+
+    if (isInsideIFrame()) {
+      this.initializeIFrameCommunication();
+    } else {
+      this.watchQueryParams();
+    }
+
+    this.listenToHighlighterEvents();
+  }
+
+  public destroy = (): void => {
+    this.unwatchQueryParams();
+    this.nodeHighlighter.destroy();
+    this.iFrameCommunicator?.destroy();
+  };
+
+  public updateConfiguration = (configuration: Partial<IPluginConfiguration>): void => {
+    this.configuration = { ...(this.configuration || defaultConfiguration), ...configuration };
+  };
+
+  private updateState = (state: Partial<IPluginState>): void => {
+    this.state = { ...this.state, ...state };
+  };
+
+  private initializeIFrameCommunication = (): void => {
+    if (!isInsideIFrame() || this.iFrameCommunicator) return;
```

Comment:
`isInsideIFrame` is already checked in constructor, I would expect the condition only at once place or this to throw in this case